lib/options_validator.js

```diff
@@ -94,6 +81,61 @@ function validate(
 
   const verifiedOptions = Object.assign({}, providedOptions);
 
+  Object.keys(optionsSchema).forEach(optionName => {
+    const optionFromSchema = optionsSchema[optionName];
+
+    if (optionFromSchema.deprecated && verifiedOptions.hasOwnProperty(optionName)) {
+      const deprecationMessage = `option [${optionName}] is deprecated and will be removed in a later version.`;
+      emitDeprecationWarning(deprecationMessage);
+    }
+
+    if (optionsValidationLevel !== VALIDATION_LEVEL_NONE) {
+      if (verifiedOptions[optionName] && optionFromSchema.type != null) {
+        if (!checkType(optionFromSchema.type, verifiedOptions[optionName])) {
+          const invalidateMessageType = `${optionName} should be of type ${
+            optionFromSchema.type
+          }, but is of type ${typeof verifiedOptions[optionName]}.`;
+          invalidate(optionsValidationLevel, invalidateMessageType, logger);
+        }
+      }
+    }
+  });
+}
+
+function checkType(requiredType, providedValue) {
+  if (typeof requiredType === 'string') {
+    if (requiredType === 'object') {
+      return isObject(providedValue);
+    } else if (requiredType === 'array') {
+      return Array.isArray(providedValue);
+    } else {
+      return typeof providedValue === requiredType;
+    }
+  } else if (Array.isArray(requiredType)) {
+    return requiredType.some(reqType => {
+      return checkType(reqType, providedValue);
+    });
+  } else if (providedValue instanceof requiredType) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * This function applies defaults without validating options.
+ * It returns an options object with defaults from optionDefaults and overrideOptions.
+ *
+ * @method
+ * @param {object} optionsSchema A schema specifying possible options and their types, defaults,
+ *   deprecation status, and whether or not they are required.
+ * @param {object} providedOptions The options passed into the operation.
+ * @param {object} optionDefaults The dynamic defaults for options in optionsSchema. If there are no dynamic defaults, null should be passed in.
+ * @param {object} [overrideOptions] The values for options that will override any user-provided value, like { readPreference: resolveReadPreference(options, { db: this.s.db, collection: this }) }.
+ * @return {object} returns an object of options with defaults set
+ */
+function applyDefaults(optionsSchema, providedOptions, optionDefaults, overrideOptions) {
+  const verifiedOptions = Object.assign({}, providedOptions);
```

Comment:
This means for all cases where we validate and apply options we do a double shallow clone. Do we need to do this?